### PR TITLE
[mediaset] Added playlist support

### PR DESCRIPTION
**Pull request by nixxo**
### Before submitting a *pull request* make sure you have:
- [ x At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Improvement

---

### Description of your *pull request* and other information

Add playlist support to the Mediaset extractor as requested in #1372

### CODE REVIEW

**Comment by nixxo** [comment]
learned something new...

File: yt_dlp/extractor/mediaset.py [hunk 1]
```diff
@@ -246,47 +248,53 @@ class MediasetShowIE(MediasetIE):
...
+    _PAGE_SIZE = 25
 
-    def _get_entries(self, title, data):
-        for entry in data or []:
-            yield self.url_result('mediaset:' + entry['guid'])
+    def _fetch_page(self, sb, page):
+        ll = page * self._PAGE_SIZE + 1
+        ul = ll + self._PAGE_SIZE - 1
+        content = self._download_json(
+            self._BY_SUBBRAND % (sb, ll, ul), sb)
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:277]: what's with these variable names? Use readable names

File: yt_dlp/extractor/mediaset.py [hunk 2]
```diff
@@ -212,3 +216,85 @@ def _real_extract(self, url):
...
+        'info_dict': {
+            'id': '100013375',
+            'title': 'I servizi',
+        },
+        'playlist_count': 53,
+    }]
+
+    _HOST = 'https://www.mediasetplay.mediaset.it'
+    _FEED_URL = 'https://feed.entertainment.tv.theplatform.eu/f/PR1GhC/'
+    _BY_SUBBRAND = _FEED_URL + 'mediaset-prod-all-programs-v2?byCustomValue={subBrandId}{%s}&sort=:publishInfo_lastPublished|desc,tvSeasonEpisodeNumber|desc&range=%d-%d'
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:270]: _BY_SUBBRAND is long, so it is justified making it a seperate variable. But the other two doesnt need to be. They are only used in a single place. Better to just put them there

File: yt_dlp/extractor/mediaset.py [hunk 3]
```diff
@@ -212,3 +216,85 @@ def _real_extract(self, url):
...
+        ll = page * self._PAGE_SIZE + 1
+        ul = ll + self._PAGE_SIZE - 1
+        content = self._download_json(
+            self._BY_SUBBRAND % (sb, ll, ul), sb)
+        if len(content.get('entries')) > 0:
+            for entry in content['entries']:
+                yield self.url_result(
+                    'mediaset:' + entry['guid'],
+                    playlist_title=entry['mediasetprogram$subBrandDescription']
+                )
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:283]:
```suggestion
                    playlist_title=entry['mediasetprogram$subBrandDescription'])
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/mediaset.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/mediaset.py:270]:
```suggestion
    _BY_SUBBRAND = 'https://feed.entertainment.tv.theplatform.eu/f/PR1GhC/mediaset-prod-all-programs-v2?byCustomValue={subBrandId}{%s}&sort=:publishInfo_lastPublished|desc,tvSeasonEpisodeNumber|desc&range=%d-%d'
```

File: yt_dlp/extractor/mediaset.py [hunk 4]
```diff
@@ -212,3 +216,85 @@ def _real_extract(self, url):
...
+    _FEED_URL = 'https://feed.entertainment.tv.theplatform.eu/f/PR1GhC/'
+    _BY_SUBBRAND = _FEED_URL + 'mediaset-prod-all-programs-v2?byCustomValue={subBrandId}{%s}&sort=:publishInfo_lastPublished|desc,tvSeasonEpisodeNumber|desc&range=%d-%d'
+    _PAGE_SIZE = 25
+
+    def _fetch_page(self, sb, page):
+        lower_limit = page * self._PAGE_SIZE + 1
+        upper_limit = lower_limit + self._PAGE_SIZE - 1
+        content = self._download_json(
+            self._BY_SUBBRAND % (sb, lower_limit, upper_limit), sb)
+        if len(content.get('entries')) > 0:
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:278]:
```suggestion
        if content.get('entries'):
```

File: yt_dlp/extractor/mediaset.py [hunk 5]
```diff
@@ -212,3 +216,85 @@ def _real_extract(self, url):
...
+                yield self.url_result(
+                    'mediaset:' + entry['guid'],
+                    playlist_title=entry['mediasetprogram$subBrandDescription']
+                )
+
+    def _real_extract(self, url):
+        playlist_id, st, sb = self._match_valid_url(url).group('id', 'st', 'sb')
+        if not sb:
+            page = self._download_webpage(url, playlist_id)
+            entries = [self.url_result(urljoin(self._HOST, url))
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:289]:
```suggestion
            entries = [self.url_result(urljoin('https://www.mediasetplay.mediaset.it', url))
```

File: yt_dlp/extractor/mediaset.py [hunk 6]
```diff
@@ -212,3 +216,82 @@ def _real_extract(self, url):
...
+    def _fetch_page(self, sb, page):
+        lower_limit = page * self._PAGE_SIZE + 1
+        upper_limit = lower_limit + self._PAGE_SIZE - 1
+        content = self._download_json(
+            self._BY_SUBBRAND % (sb, lower_limit, upper_limit), sb)
+        if content.get('entries'):
+            for entry in content['entries']:
+                yield self.url_result(
+                    'mediaset:' + entry['guid'],
+                    playlist_title=entry['mediasetprogram$subBrandDescription'])
```
Comment (pukkandan) [inline yt_dlp/extractor/mediaset.py:280]:
```suggestion
        for entry in content.get('entries') or []:
            yield self.url_result(
                'mediaset:' + entry['guid'],
                playlist_title=entry['mediasetprogram$subBrandDescription'])
```

**Review by pukkandan — APPROVED** [review]

**Comment by nixxo** [comment]
yes, everything is ok

**Comment by Heros99** [comment]
debug: None of the available releases match the specified AssetType, ProtectionScheme, and/or Format preferences.
is it possible to add it to KnownDRMIE or change apikey?
Thank you very much

**Comment by nixxo** [comment]
Can you provide a url?

**Comment by Heros99** [comment]
> Can you provide a url?

https://mediasetinfinity.mediaset.it/video/leduefaccedellamore/episodio-1-e-lamore-che-ti-sceglie_F302632502000101

**Comment by nixxo** [comment]
ok, this will require a bit of rewriting of the extractor because Mediaset put everything in mediasetinfinity and there are free content and payed content on the same website and it requires some extra better drm/login detection. I'll look into it.

**Comment by Heros99** [comment]
> ok, this will require a bit of rewriting of the extractor because Mediaset put everything in mediasetinfinity and there are free content and payed content on the same website and it requires some extra better drm/login detection. I'll look into it.

Thank you very much

**Comment by Heros99** [comment]
C:\Users\******\Desktop\yt-dlp download>yt-dlp.exe -Uv --cookies-from-browser firefox https://mediasetinfinity.mediaset.it/video/leduefaccedellamore/episodio-1-e-lamore-che-ti-sceglie_F302632502000101
[debug] Command-line config: ['-Uv', '--cookies-from-browser', 'firefox', 'https://mediasetinfinity.mediaset.it/video/leduefaccedellamore/episodio-1-e-lamore-che-ti-sceglie_F302632502000101']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] ytdl-patched/yt-dlp version 2022.11.29.40298 [fd90e3d] (win_exe)
[debug] ** This build is unofficial daily builds, provided for ease of use.
[debug] ** Please do not ask for any support.
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19044-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg N-106564-gd6d46a2c50-20220411 (setts), ffprobe N-106564-gd6d46a2c50-20220411
[debug] Optional libraries: Cryptodome-3.16.0, brotli-1.0.9, certifi-2022.09.24, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "C:\Users\******\AppData\Roaming\Mozilla\Firefox\Profiles\*******.default-release\cookies.sqlite"
[Cookies] Extracted 2531 cookies from firefox
[debug] Proxy map: {}
[debug] Loaded 1730 extractors
[debug] Fetching release info: https://api.github.com/repos/ytdl-patched/yt-dlp/releases/latest
Latest version: 2022.11.29.40298, Current version: 2022.11.29.40298
yt-dlp is up to date (2022.11.29.40298)
[debug] [Mediaset] Extracting URL: https://mediasetinfinity.mediaset.it/video/leduefaccedellamore/episodio-1-e-lamore-che-ti-sceglie_F302632502000101
[Mediaset] F302632502000101: Downloading JSON metadata
[Mediaset] F302632502000101: Downloading MPEG4 SMIL data
[Mediaset] F302632502000101: Downloading M3U SMIL data
ERROR: [Mediaset] F302632502000101: None of the available releases match the specified AssetType, ProtectionScheme, and/or Format preferences
  File "yt_dlp\extractor\common.py", line 678, in extract
  File "yt_dlp\extractor\mediaset.py", line 248, in _real_extract
  File "yt_dlp\extractor\mediaset.py", line 230, in _real_extract
  File "yt_dlp\extractor\theplatform.py", line 45, in _extract_theplatform_smil

**always the same**

**Comment by nixxo** [comment]
@Heros99 you are not using the version with the improvements... 

```
❯ ydpm https://mediasetinfinity.mediaset.it/video/leduefaccedellamore/episodio-1-e-lamore-che-ti-sceglie_F302632502000101 -F
[Mediaset] F302632502000101: Downloading JSON metadata
[Mediaset] F302632502000101: Downloading MPEG4 SMIL data
[Mediaset] F302632502000101: Downloading MPEG-DASH SMIL data
[Mediaset] F302632502000101: Downloading M3U SMIL data
ERROR: [Mediaset] F302632502000101: Content behind paywall and DRM
```

**Pull request merged**